File: android/rctmln/src/main/java/com/maplibre/rctmln/components/camera/CameraUpdateQueue.java
package com.maplibre.rctmln.components.camera;

import com.mapbox.mapboxsdk.maps.MapboxMap;
import com.maplibre.rctmln.components.mapview.RCTMLNMapView;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.FutureTask;


/**
 * Created by nickitaliano on 9/5/17.
 */

public class CameraUpdateQueue {
    private Queue<CameraStop> mQueue;

    private OnCompleteAllListener mCompleteListener;

    public interface OnCompleteAllListener {
        void onCompleteAll();
    }

    public CameraUpdateQueue() {
        mQueue = new LinkedList<>();
    }

    public void offer(CameraStop item) {
        mQueue.offer(item);
    }

    public int size() {
        return mQueue.size();
    }

    public boolean isEmpty() {
        return mQueue.isEmpty();
    }

    public void flush() {
        while (mQueue.size() > 0) {
            mQueue.remove();
        }
        mQueue = new LinkedList<>();
    }

    public void setOnCompleteAllListener(OnCompleteAllListener listener) {
        mCompleteListener = listener;
    }

    public void execute(RCTMLNMapView map) {
        if (mQueue.isEmpty()) {
            if (mCompleteListener != null) {
                mCompleteListener.onCompleteAll();
            }
            return;
        }

        final CameraStop stop = mQueue.poll();
        if (stop == null) {
            return;
        }

        CameraUpdateItem item = stop.toCameraUpdate(map);
        item.run();
        execute(map);
    }
}

File: package/android/src/main/java/org/maplibre/reactnative/components/camera/CameraUpdateQueue.kt
package org.maplibre.reactnative.components.camera

import org.maplibre.reactnative.components.mapview.MLRNMapView
import java.util.LinkedList
import java.util.Queue

class CameraUpdateQueue {
    private var queue: Queue<CameraStop?>

    private var completeListener: OnCompleteAllListener? = null

    interface OnCompleteAllListener {
        fun onCompleteAll()
    }

    init {
        queue = LinkedList<CameraStop?>()
    }

    fun offer(item: CameraStop?) {
        queue.offer(item)
    }

    fun size(): Int = queue.size

    val isEmpty: Boolean
        get() = queue.isEmpty()

    fun flush() {
        while (queue.size > 0) {
            queue.remove()
        }
        queue = LinkedList<CameraStop?>()
    }

    fun setOnCompleteAllListener(listener: OnCompleteAllListener?) {
        completeListener = listener
    }

    fun execute(map: MLRNMapView) {
        if (queue.isEmpty()) {
            if (completeListener != null) {
                completeListener!!.onCompleteAll()
            }

            return
        }

        val stop = queue.poll()
        if (stop == null) {
            return
        }

        val item = stop.toCameraUpdate(map)
        item.run()
        execute(map)
    }
}
